access/handler.go

```diff
@@ -701,6 +725,392 @@ func (h *Handler) GetExecutionResultByID(ctx context.Context, req *access.GetExe
+// - codes.InvalidArgument - if unknown block status provided.
+// - codes.ResourceExhausted - if the maximum number of streams is reached.
+// - codes.Internal - if stream encountered an error, if stream got unexpected response or could not convert block to message or could not send response.
+func (h *Handler) SubscribeBlocksFromStartHeight(request *access.SubscribeBlocksFromStartHeightRequest, stream access.AccessAPI_SubscribeBlocksFromStartHeightServer) error {
+	// check if the maximum number of streams is reached
+	if h.StreamCount.Load() >= h.MaxStreams {
+		return status.Errorf(codes.ResourceExhausted, "maximum number of streams reached")
+	}
+	h.StreamCount.Add(1)
+	defer h.StreamCount.Add(-1)
+
+	blockStatus, err := h.getBlockStatus(request.GetBlockStatus())
+	if err != nil {
+		return err
+	}
+
+	sub := h.api.SubscribeBlocks(stream.Context(), flow.ZeroID, request.GetStartBlockHeight(), blockStatus)
+	return h.handleBlocksSubscription(sub, stream.Send, request.GetFullBlockResponse(), blockStatus)
+}
+
+// SubscribeBlocksFromLatest handles subscription requests for blocks started from latest sealed block.
+// It takes a SubscribeBlocksFromLatestRequest and an AccessAPI_SubscribeBlocksFromLatestServer stream as input.
+// The handler manages the subscription to block updates and sends the subscribed block information
+// to the client via the provided stream.
+//
+// Expected errors during normal operation:
+// - codes.InvalidArgument - if unknown block status provided.
+// - codes.ResourceExhausted - if the maximum number of streams is reached.
+// - codes.Internal - if stream encountered an error, if stream got unexpected response or could not convert block to message or could not send response.
+func (h *Handler) SubscribeBlocksFromLatest(request *access.SubscribeBlocksFromLatestRequest, stream access.AccessAPI_SubscribeBlocksFromLatestServer) error {
+	// check if the maximum number of streams is reached
+	if h.StreamCount.Load() >= h.MaxStreams {
+		return status.Errorf(codes.ResourceExhausted, "maximum number of streams reached")
+	}
+	h.StreamCount.Add(1)
+	defer h.StreamCount.Add(-1)
+
+	blockStatus, err := h.getBlockStatus(request.GetBlockStatus())
+	if err != nil {
+		return err
+	}
+
+	sub := h.api.SubscribeBlocks(stream.Context(), flow.ZeroID, 0, blockStatus)
+	return h.handleBlocksSubscription(sub, stream.Send, request.GetFullBlockResponse(), blockStatus)
+}
+
+// handleBlocksSubscription handles the subscription to block updates and sends
+// the subscribed block information to the client via the provided stream.
+//
+// Expected errors during normal operation:
+//   - codes.Internal: if the stream encounters an error, gets an unexpected response,
+//     or cannot convert block to message or send a response.
+func (h *Handler) handleBlocksSubscription(sub subscription.Subscription, send sendSubscribeBlocksResponseFunc, fullBlockResponse bool, blockStatus flow.BlockStatus) error {
```

Comment:
you can reduce a bit of boilerplate by doing something like this

```golang
func (h *Handler) handleBlocksResponse(stream access.AccessAPI_SubscribeBlocksFromLatestServer, fullBlockResponse bool, blockStatus flow.BlockStatus) func(*flow.Block) error {
	return func(block *flow.Block) error {
		msgBlockResponse, err := h.blockResponse(block, fullBlockResponse, blockStatus)
		if err != nil {
			return rpc.ConvertError(err, "could not convert block to message", codes.Internal)
		}

		err = stream.Send(&access.SubscribeBlocksResponse{
			Block: msgBlockResponse.Block,
		})
		if err != nil {
			return rpc.ConvertError(err, "could not send response", codes.Internal)
		}

		return nil
	}
}

func handleSubscription[T any](sub subscription.Subscription, handleResponse func(block T) error) error {
	for {
		v, ok := <-sub.Channel()
		if !ok {
			if sub.Err() != nil {
				return rpc.ConvertError(sub.Err(), "stream encountered an error", codes.Internal)
			}
			return nil
		}

		block, ok := v.(T)
		if !ok {
			return status.Errorf(codes.Internal, "unexpected response type: %T", v)
		}

		return handleResponse(block)
	}
}
```


then the function that's created for each type only has its unique code.

eventually, we should update subscriptions to use generics, but we can save that for another day.